src/lang/modifyAst/addFillet.ts

```diff
@@ -64,120 +190,101 @@ export function addFillet(
-    (extrudeInit.type !== 'CallExpression' &&
-      extrudeInit.type !== 'PipeExpression')
+    extrudeInit.type !== 'CallExpression' &&
+    extrudeInit.type !== 'PipeExpression'
   ) {
-    return new Error('Extrude PipeExpression / CallExpression not found.')
+    return new Error('Extrude must be a PipeExpression or CallExpression')
   }
 
-  // determine if extrude is in a PipeExpression or CallExpression
+  const extrudeExpressionType = extrudeInit.type
 
+  return { extrudeDeclarator, extrudeInit }
+}
+
+function handleExtrudeType(
+  node: Program,
+  extrudeDeclarator: VariableDeclarator,
+  extrudeInit: Expr,
+  filletCall: CallExpression,
+  pathToExtrudeNode: PathToNode,
+  tag: string
+): { modifiedAst: Program; pathToFilletNode: PathToNode } | Error {
   // CallExpression - no fillet
   // PipeExpression - fillet exists
 
-  const getPathToNodeOfFilletLiteral = (
-    pathToExtrudeNode: PathToNode,
-    extrudeDeclarator: VariableDeclarator,
-    tag: string
-  ): PathToNode => {
-    let pathToFilletObj: any
-    let inFillet = false
-    traverse(extrudeDeclarator.init, {
-      enter(node, path) {
-        if (node.type === 'CallExpression' && node.callee.name === 'fillet') {
-          inFillet = true
-        }
-        if (inFillet && node.type === 'ObjectExpression') {
-          const hasTag = node.properties.some((prop) => {
-            const isTagProp = prop.key.name === 'tags'
-            if (isTagProp && prop.value.type === 'ArrayExpression') {
-              return prop.value.elements.some(
-                (element) =>
-                  element.type === 'Identifier' && element.name === tag
-              )
-            }
-            return false
-          })
-          if (!hasTag) return false
-          pathToFilletObj = path
-          node.properties.forEach((prop, index) => {
-            if (prop.key.name === 'radius') {
-              pathToFilletObj.push(
-                ['properties', 'ObjectExpression'],
-                [index, 'index'],
-                ['value', 'Property']
-              )
-            }
-          })
-        }
-      },
-      leave(node) {
-        if (node.type === 'CallExpression' && node.callee.name === 'fillet') {
-          inFillet = false
-        }
-      },
-    })
-    let indexOfPipeExpression = pathToExtrudeNode.findIndex(
-      (path) => path[1] === 'PipeExpression'
-    )
-    indexOfPipeExpression =
-      indexOfPipeExpression === -1
-        ? pathToExtrudeNode.length
-        : indexOfPipeExpression
-
-    return [
-      ...pathToExtrudeNode.slice(0, indexOfPipeExpression),
-      ...pathToFilletObj,
-    ]
-  }
-
   if (extrudeInit.type === 'CallExpression') {
     // 1. no fillet case
     extrudeDeclarator.init = createPipeExpression([extrudeInit, filletCall])
     return {
-      modifiedAst: _node,
+      modifiedAst: node,
       pathToFilletNode: getPathToNodeOfFilletLiteral(
         pathToExtrudeNode,
         extrudeDeclarator,
```

Comment:
This function is getting a bit confusing because you need to know how mutating structures effect each other.

so here we mutate one object `extrudeDeclarator`
`extrudeDeclarator.init = createPipeExpression([extrudeInit, filletCall])`
And then immediately return another object `node`
`return { modifiedAst: node, ...`
And the reason this works is because `extrudeDeclarator` is actually a child/part of `node` so by mutating the former we are also mutating the later. The problem with this is that's completely opaque to someone reading and trying to understand this function for the first time, and in fact the function won't work as intended if for some reason someone else used this function and passed in  a `extrudeDeclarator` that was not actually part of `node` (this isn't an exported function so that's unlikely to happen, but the readability still stands)


 I think breaking this all into so many functions here is a bit of a mixed blessing, while lots of functions can help readability from the sense of getting the narrative of what's happening, it does abstract you away from the nuts and bolt and when the nuts and bolt is mutating two variables that are actually implicitly connected it's a problem.
 
In cases of ast mods, queries can be broken out in helper functions, but mutations probably need to stay in the one function so that the whole story is contained in that function.

My apologies because I know I pointing this out now, when some of these abstractions were done in previous PRs.

Does what I'm saying make sense?

Reply:
Been thinking about this comment a bit more, I don't like that its so high level without concrete examples of what I mean particularly:

> In cases of ast mods, queries can be broken out in helper functions, but mutations probably need to stay in the one function so that the whole story is contained in that function.

I think this is an hypothetical function that I think mixes well both adding abstractions while none of the mutations leave the main function:

```ts
function hyotheticalDoStuff(
  ast: Program,
  pathToNode1: PathToNode,
  pathToNode2: PathToNode,
): {modifiedAst: Program} | Error {
  const astClone = structuredClone(ast)

  const node1 = getNodeFromPath<CallExpression>(astClone, pathToNode1, 'CallExpression')
  const node2 = getNodeFromPath<VariableDeclaration>(astClone, pathToNode2, 'VariableDeclaration')

  if(err(node1)) return node1
  if(err(node2)) return node2

  // any bussiness logic checks for these nodes
  const isNode1Good = checkNode1Specifics(node1)
  const isNode1Good = checkNode2Specifics(node1)

  if (err(isNode1Good)) return isNode1Good
  if (!isNode1Good) return new Error('Node 1 is not good')

  // lets say we need to update the callexpressions arguments and the variable declarations init
  // we have a helper function that creates the updated node, but does not mutate anything
  const updatedChildeNodePart1 = updateChildNode1(node1.node)
  const updatedChildeNodePart2 = updateChildNode2(node1.node)

  // we wait till we're back to the main thread to apply the changes
  node1.node.arguments[0] = updatedChildeNodePart1
  node2.node.declarations[0].init = updatedChildeNodePart2

  // this way when reading this function we can read the "story" that node1 and node2 are children that come from the astClone object
  // and that by mutating them we're also mutating the astClone object and that's why we're returning astClone.
  // none of the mutations are tucked away in the helper functions, if they were then it would be using an input param as the output, which we want to avoid.

  return {modifiedAst: astClone}
}
```

I really hope I'm making sense with all this.

Reply:
this is the coolest comment ever, I'll grab a ~~beer~~ cup of tea and read it a couple times more.

Reply:
My goal was to break down the complex logic into simple pieces—`addFillet` tags the segment and adds the fillet, and the helpers handle the details. 

Complex shit is just piles of simple tasks, so I organized them that way.

But I see your point about keeping AST mutations together for clarity. Yeah, I totally missed the moment that mutations are not obvious.

Reply:
Thank you, and thanks for linking to the commit.

I'm not sure what you mean by `createPipeExpression` because that function does not mutate anything, it's creating an entirely new data structure.

I should clarify, I'm using mutate here to mean changing underlying memory.

like this would be a function that mutates
```js
const doesMutate = (val) => {
  val.newProperty = 'hi'
  return val
}
```
Where is this does the same thing, but because it does a shallow close, not not effect the underlying memory, because it creates a new object with the `...` spread operator
```js
const doesNotMutate = (val) => {
  const newVal = {...val, newProperty: 'hi}
  return newVal
}
```
I might as well be thorough at risk of making you read stuff you're already aware of.

This relates to much of the idea of "pure function", as a pure function is not supported to effect anything outside of itself (that is no "side effects") and mutating is a side effect because it can effect other parts of the code and lead to tricky bugs.

The thing I was most concerned about that I called out last time was mutating extrudeDeclarator in order to also mutate the ast because `extrudeDeclarator` was a child of the ast, the connection between the two was implicit and very hard to follow without going outside the function and looking at how it was being called. Here's another example of functions params being mutated causing issues, 

![image](https://github.com/user-attachments/assets/5cfc11c9-c73d-4942-b0ce-c964cee5f99c)

I've intentionally cut some of setup code from the start of this snippet for the purposes of putting yourself in the shoes of someone trying to understand what the code is doing, I can imagine thinking to myself

> "why did `itGets` log out two different things, when I called it with the exact same pararm? sure I know that `itMutates` mutates, but they use different variables, `itGets` using `yo` and it mutates uses `myAbc`!?"

The full snipped explains it because `yo` and `myAbc` in infact connected, so by muting `myAbc` we're also mutating `yo`

![image](https://github.com/user-attachments/assets/daabf11f-f549-4efc-ab60-37e7b9456979)

In this toy example seeing the connection is very clear because it's just above, but it's get tricky of those two variables get passed along through multiple function, you have to go upstream to realise there's two separate vars are linked.

I think maybe I should clarify what a mutation looks like javascript
for objects if you see `myObj.propertyName =` or `myObj[stringVariable] =` than `myObj` is being mutated
For arrays it's probably more like `.push(), .pop(), .shift(), .unshift()`

Maybe that relates to your comment about `createPipeExpression`, because there's the line `extrudeDeclarator.init = createPipeExpression([...`?

So there's a mutation here but it's not `createPipeExpression`, it's the first half `extrudeDeclarator.init =` where the mutation is happening, but since your change in https://github.com/KittyCAD/modeling-app/commit/045a05ae1086ff5b215240ea3c6eac44327cfafd that mutation is now happening on a variable that's declared in the same function so it's much easier to follow, where as before it was being passed in as an input param.

Reply:
I think I need to develop a habit of printing out your comments and hanging them on the wall. in the frame :D thank you!